client/src/controllers/LoaderController.ts

```diff
@@ -0,0 +1,92 @@
+import { Controller } from '@hotwired/stimulus';
+
+/**
+ * <button
+ *   type="submit"
+ *   class="button button-longrunning"
+ *   data-controller="w-loader"
+ *   data-w-loader-active-class="button-longrunning-active"
+ *   data-w-loader-active-value="{% trans 'Signing in…' %}"
+ *   data-w-loader-duration-seconds-value="40"
+ *   data-action="w-loader#activate"
+ * >
+ *  {% icon name="spinner" %}
+ *  <em data-w-loader-target="label">{% trans 'Sign in' %}</em>
+ * </button>
+ */
+export class LoaderController extends Controller {
+  static classes = ['active'];
+  static targets = ['label'];
+  static values = {
+    active: { type: String, default: 'Loading' },
+    durationSeconds: { type: Number, default: 30 },
+  };
+
+  activeClass: string;
+  activeValue: string;
+  durationSecondsValue: number;
+  labelTarget: HTMLElement;
+  hasLabelTarget: boolean;
+
+  activate() {
+    let originalText;
+    if (this.hasLabelTarget === true) {
+      originalText = this.labelTarget.innerText;
+    }
+    originalText = this.element.textContent;
+
+    this.cancelSpinner(originalText);
+
+    // If client-side validation is active on this form, and is going to block submission of the
+    // form, don't activate the spinner
+    const form = this.element.closest('form');
+
+    if (
+      form &&
+      form.checkValidity &&
+      !form.noValidate &&
+      !form.checkValidity()
+    ) {
+      return;
+    }
+
+    // set time out
+    this.setTimeout(originalText);
+  }
+
+  cancelSpinner(originalText: string) {
+    this.element.removeAttribute('disabled');
+    this.element.classList.remove(this.activeClass);
+
+    if (this.activeValue && this.hasLabelTarget === true) {
+      this.labelTarget.innerText = originalText;
+    }
+  }
+
+  setTimeout(originalText: string) {
+    const timeout = setTimeout(() => {
+      const reEnableAfter = 30;
+
+      if (!this.durationSecondsValue) {
+        // Button re-enables after a timeout to prevent button becoming
+        // permanently un-usable
+        this.durationSecondsValue = setTimeout(() => {
+          clearTimeout(this.durationSecondsValue);
+
+          this.cancelSpinner(originalText);
+        }, reEnableAfter * 1000);
+      }
+
+      if (this.activeValue && this.hasLabelTarget === true) {
+        // Sets the targetElement innerText to the activeValue
+        this.labelTarget.innerText = this.activeValue;
+      }
+      // Disabling button must be done last: disabled buttons can't be
+      // modified in the normal way, it would seem.
+      this.element.classList.add(this.activeClass);
```

Comment:
```suggestion
      this.element.classList.add(this.hasActiveClass ? this.activeClass : 'button-longrunning-active');
```

To aid in the transition to this new approach, maybe we should support a default value here, we could even store this as a static property on the class (we can discuss this). It would mean that most of the time we would not need to put this value into the HTML.